src/main/java/org/openjdk/jextract/impl/TreeMaker.java

```diff
@@ -103,18 +103,25 @@ public Declaration createTree(Cursor c) {
     }
 
     private Declaration createTreeInternal(Cursor c) {
-        return switch (c.kind()) {
-            case EnumDecl -> createEnum(c);
-            case EnumConstantDecl -> createEnumConstant(c);
-            case FieldDecl -> createVar(c, Declaration.Variable.Kind.FIELD);
-            case ParmDecl -> createVar(c, Declaration.Variable.Kind.PARAMETER);
-            case FunctionDecl -> createFunction(c);
-            case StructDecl -> createRecord(c, Declaration.Scoped.Kind.STRUCT);
-            case UnionDecl -> createRecord(c, Declaration.Scoped.Kind.UNION);
-            case TypedefDecl -> createTypedef(c);
-            case VarDecl -> createVar(c, Declaration.Variable.Kind.GLOBAL);
-            default -> null; // skip
-        };
+        Position pos = CursorPosition.of(c);
+        // dedup multiple declarations that point to the same source location
+        Declaration decl = declarationCache.get(pos);
+        if (decl == null) {
+            decl = switch (c.kind()) {
+                case EnumDecl -> createEnum(c);
+                case EnumConstantDecl -> createEnumConstant(c);
+                case FieldDecl -> createVar(c, Declaration.Variable.Kind.FIELD);
+                case ParmDecl -> createVar(c, Declaration.Variable.Kind.PARAMETER);
+                case FunctionDecl -> createFunction(c);
+                case StructDecl -> createRecord(c, Declaration.Scoped.Kind.STRUCT);
+                case UnionDecl -> createRecord(c, Declaration.Scoped.Kind.UNION);
+                case TypedefDecl -> createTypedef(c);
+                case VarDecl -> createVar(c, Declaration.Variable.Kind.GLOBAL);
+                default -> null; // skip
+            };
+            declarationCache.put(pos, decl);
+        }
+        return decl;
```

Comment:
IIRC we had a cache here before, but removed it because it was faster to re-create trees. That might have inadvertently also created latent issues with inconsistent IR views of the same decl?

Any way, we need the cache to 'canonicalize' IR nodes. Good to know.

Reply:
Yes, we used to have a cache and removed it - I don't think we thought too much about what it meant in terms of losing indentity of the objects involved, since we never relied on that anyway. But I remember this being a problem when we took our first stab at the name mangling patch, which was confirmed with an experiment I did few days ago: w/o de-duplication, it is possible to set an attribute on a struct, but to still have types in the IR that point to the old, un-attributes struct, which is bad.